File: src/Mahal.c
/*************************************************************************
 
 (c) 2008-2020 Guillaume Guénard
 Université de Montréal, Montreal, Quebec, Canada
 
 **Mahalanobis Distance**
 
 This file is part of codep
 
 codep is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 codep is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with codep.  If not, see <https://www.gnu.org/licenses/>.
 
 C fonctions definitions
 
 *************************************************************************/

#include"Mahal.h"

/* Computes the Mahalanobis distance between two matrices.
 That function can take two matrices with identical numbers
 of columns and output a rectangular
 distance matrix.
 */

void dist_Mahal(double* from, double* to, int* n, int* tri, int* m,
                double* d,
                double* cov, int* sq) {
  int end, i, j, k, l, ii, jj;
  double acc;
  end = (*tri) ? n[0]*(n[0]-1)/2 : n[0]*n[1];
  for(i = 0, j = (*tri) ? 1 : 0, k = 0; k < end; k++) {
    for(l = 0, ii = i, jj = j; l < *m; l++,
        ii += n[0], jj += n[1]) {
      //
      // mod from here...
      acc = from[ii] - to[jj];
      acc *= acc;
      d[k] += acc;
      // mod to there.
      //
    }
    if(!*sq)
      d[k] = sqrt(d[k]);
    j++;
    if(j == n[1]) {
      i++;
      j = (*tri) ? i+1 : 0;
    }
    if(!(k%INTERRUPT_CHECK))
      R_CheckUserInterrupt();
  }
  return;
}

File: Archive/R/Mahal.R
## **************************************************************************
##
##    (c) 2018-2022 Guillaume Guénard
##        Department de sciences biologiques,
##        Université de Montréal
##        Montreal, QC, Canada
##
##    **Mahalanobis distance between two point sets A and B**
##
##    This file is part of codep
##
##    codep is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    codep is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with codep. If not, see <https://www.gnu.org/licenses/>.
##
##    R source code file
##
## **************************************************************************
##
#' Calculation of the Mahalanobis Distance
#' 
#' Function \code{Mahal} carries out the calculation of pairwise Mahalanobis
#' distances within a set of coordinates or between two sets thereof, with
#' optional weights.
#' 
#' @param x A set of coordinates in the form of a \code{\link{matrix}} or
#' \code{\link{data.frame}}.
#' @param y An optional second set of coordinates in the same dimensions as
#' argument \code{x}.
#' @param cov The covariance (or inverse covariance) matrix of the distribution.
#' @param squared Should the squared Euclidean distances be returned (default:
#' FALSE).
#' @param inverted A \code{logical} specifying whether \code{cov} contains the
#' inverse of the covariance matrix (default: \code{FALSE}).
#' @param ... Arguments to be passed to \code{solve} for computing the inverse
#' of the covariance matrix (if \code{inverted == FALSE}).
#' 
#' @return A `\link{dist}-class` object or, whenever \code{y} is provided,
#' a \code{\link{matrix}} with as many rows as the number of rows in \code{x}
#' and as many columns as the number of rows in \code{y}.
#' 
#' @details When only one set of coordinates is given to the function (i.e.,
#' when argument \code{y} is omitted), the function returns the pairwise
#' distances in the form of a `\link{dist}-class` object representing a
#' lower-triangle matrix. If weights are omitted, the result is identical to
#' that produced by function \link{dist} with argument
#' \code{method = "mahalanobis"} (the function's default).
#' 
#' The standard `R` function used to calculate distances (\code{\link{dist}}),
#' only allows one to calculate pairwise distances between the rows of a single
#' matrix of Cartesian coordinates and return a `\link{dist}-class` object,
#' which is a one-dimensional array meant to be interpreted as a
#' lower-triangular matrix. Function \code{Mahal} can also be provided two data
#' matrices (arguments \code{x} and \code{y}) and output a rectangular matrix of
#' the Euclidean distances. Additionally, the function can be used to calculate
#' the weighted Euclidean distance (which if a particular case of the
#' Mahalanobis distance) by providing cov with a vector of weights (or inverse
#' weights).
#' 
#' @author \packageAuthor{codep}
#' Maintainer: \packageMaintainer{codep}
#' 
#' @seealso The `\link{dist}-class` and associated methods.
#' 
#' @examples
#' 
#' @useDynLib codep, .registration = TRUE
#' 
#' @export
Mahal <- function(x, y, cov, squared = FALSE, inverse = FALSE, ...) {
  m <- NCOL(x)
  if (!is.matrix(x))
    x <- as.matrix(x)
  storage.mode(x) <- "double"
  if(!is.matrix(cov))
    stop("Argument 'cov' must be a matrix!")
  if(nrow(cov) != ncol(cov))
    stop("Argument 'cov' must be a square matrix!")
  if(!inverse)
    cov <- solve(cov, ...)
  storage.mode(cov) <- "double"
  if (!missing(y)) {
    if(NCOL(y) != m)
      stop("'y' must have the same number of coordinates as 'x'!")
    if(!is.matrix(y))
      y <- as.matrix(y)
    storage.mode(y) <- "double"
    N <- c(NROW(x), NROW(y))
    out <- matrix(0, N[2L], N[1L])
    rownames(out) <- dimnames(y)[[1L]]
    colnames(out) <- dimnames(x)[[1L]]
  } else {
    N <- c(NROW(x),NROW(x))
    out <- double(N[1L]*(N[1L] - 1L)/2)
  }
  res <- .C(
    "dist_Mahal",
    x,
    if(missing(y)) x else y,
    N,
    missing(y),
    m,
    out,
    cov,
    squared
  )
  out[] <- res[[6L]]
  attr(out, "method") <- "mahalanobis"
  attr(out, "call") <- match.call()
  if (missing(y)) {
    attr(out, "Size") <- N[1L]
    attr(out, "Labels") <- dimnames(x)[[1L]]
    attr(out, "Diag") <- FALSE
    attr(out, "Upper") <- FALSE
    attr(out, "class") <- "dist"
  }
  return(out)
}
#'
